File: src/utils/scheduling.js
import { addDays, computeEndDate, rangesOverlap, formatDate } from './dates.js';

/**
 * Check if a given equipment is booked during the specified date range.
 * Returns the conflicting booking(s) or an empty array.
 *
 * @param {Array} allBookings - flat array of all bookings across all projects
 * @param {string} equipmentId
 * @param {string} startDate - YYYY-MM-DD
 * @param {string} endDate - YYYY-MM-DD (inclusive)
 * @param {string|null} excludeBookingId - optional booking id to exclude (for edits)
 * @returns {Array} conflicting bookings
 */
export function checkConflict(allBookings, equipmentId, startDate, endDate, excludeBookingId = null) {
  return allBookings.filter((b) => {
    if (b.equipmentId !== equipmentId) return false;
    if (excludeBookingId && b.id === excludeBookingId) return false;
    return rangesOverlap(startDate, endDate, b.startDate, b.endDate);
  });
}

/**
 * Find the next available start date for an equipment item,
 * starting from requestedStart, for a given duration.
 *
 * @param {Array} allBookings
 * @param {string} equipmentId
 * @param {string} requestedStart - YYYY-MM-DD
 * @param {number} durationDays
 * @param {string|null} excludeBookingId
 * @returns {string} next available start date (YYYY-MM-DD)
 */
export function findNextAvailableDate(allBookings, equipmentId, requestedStart, durationDays, excludeBookingId = null) {
  let candidate = requestedStart;
  let iterations = 0;
  const maxIterations = 365 * 2; // safety cap

  while (iterations < maxIterations) {
    const candidateEnd = computeEndDate(candidate, durationDays);
    const conflicts = checkConflict(allBookings, equipmentId, candidate, candidateEnd, excludeBookingId);

    if (conflicts.length === 0) {
      return candidate;
    }

    // Move candidate to the day after the latest conflict end
    const latestEnd = conflicts.reduce((max, b) => (b.endDate > max ? b.endDate : max), '');
    candidate = addDays(latestEnd, 1);
    iterations++;
  }

  return candidate;
}

/**
 * Collect all bookings from a projects array into a flat array.
 */
export function flattenBookings(projects) {
  const bookings = [];
  for (const project of projects) {
    for (const deliverable of (project.deliverables || [])) {
      for (const booking of (deliverable.bookings || [])) {
        bookings.push({ ...booking, projectId: project.id, projectName: project.name });
      }
    }
  }
  return bookings;
}

/**
 * Given a list of projects and equipment list, return bookings enriched with
 * project color, equipment name, etc.
 */
export const PROJECT_COLORS = [
  '#0d9488', // teal-600
  '#7c3aed', // violet-600
  '#dc2626', // red-600
  '#d97706', // amber-600
  '#2563eb', // blue-600
  '#16a34a', // green-600
  '#db2777', // pink-600
  '#ea580c', // orange-600
  '#0891b2', // cyan-600
  '#65a30d', // lime-600
];

export function getProjectColor(projectId, projects) {
  const idx = projects.findIndex((p) => p.id === projectId);
  return PROJECT_COLORS[idx % PROJECT_COLORS.length] || PROJECT_COLORS[0];
}

/**
 * Compute total cost for a set of bookings given equipment map.
 */
export function computeTotalCost(bookings, equipmentMap) {
  return bookings.reduce((total, b) => {
    const eq = equipmentMap[b.equipmentId];
    if (!eq) return total;
    return total + eq.costPerDay * b.durationDays;
  }, 0);
}

/**
 * Generate a unique ID with a given prefix.
 */
export function generateId(prefix = 'id') {
  return `${prefix}-${Date.now()}-${Math.random().toString(36).slice(2, 7)}`;
}

/**
 * Get bookings upcoming within the next N days from today.
 */
export function getUpcomingBookings(allBookings, days = 7) {
  const now = formatDate(new Date());
  const future = addDays(now, days);
  return allBookings.filter((b) => b.startDate >= now && b.startDate <= future);
}

File: streamlit_app/utils/scheduling.py
from datetime import datetime, timedelta


def str_to_date(s):
    return datetime.strptime(s, "%Y-%m-%d").date()


def date_to_str(d):
    return d.strftime("%Y-%m-%d")


def compute_end_date(start_str, duration_days):
    start = str_to_date(start_str)
    end = start + timedelta(days=duration_days - 1)
    return date_to_str(end)


def ranges_overlap(a_start, a_end, b_start, b_end):
    return a_start <= b_end and a_end >= b_start


def flatten_bookings(projects):
    """Return flat list of all bookings with project_name and deliverable_name injected."""
    bookings = []
    for p in projects:
        for d in p.get("deliverables", []):
            for b in d.get("bookings", []):
                bookings.append({
                    **b,
                    "project_name": p["name"],
                    "deliverable_name": d["name"],
                })
    return bookings


def check_conflict(all_bookings, equipment_id, start_date, end_date, exclude_id=None):
    """Return list of conflicting bookings."""
    conflicts = []
    for b in all_bookings:
        if b["equipment_id"] != equipment_id:
            continue
        if exclude_id and b["id"] == exclude_id:
            continue
        if ranges_overlap(start_date, end_date, b["start_date"], b["end_date"]):
            conflicts.append(b)
    return conflicts


def find_next_available(all_bookings, equipment_id, requested_start, duration_days, exclude_id=None):
    """Find next available start date with no conflicts.

    Returns (start_date_str, was_rescheduled).
    """
    candidate = requested_start
    for _ in range(730):  # max 2 years
        end = compute_end_date(candidate, duration_days)
        conflicts = check_conflict(all_bookings, equipment_id, candidate, end, exclude_id)
        if not conflicts:
            return candidate, candidate != requested_start
        # Jump to day after latest conflict end
        latest_end = max(c["end_date"] for c in conflicts)
        next_day = str_to_date(latest_end) + timedelta(days=1)
        candidate = date_to_str(next_day)
    return candidate, True


def compute_project_cost(project, equipment_map):
    """Compute total cost for a project."""
    total = 0
    for d in project.get("deliverables", []):
        for b in d.get("bookings", []):
            eq = equipment_map.get(b["equipment_id"])
            if eq:
                total += eq["cost_per_day"] * b["duration_days"]
    return total


def compute_deliverable_cost(deliverable, equipment_map):
    """Compute total cost for a deliverable."""
    total = 0
    for b in deliverable.get("bookings", []):
        eq = equipment_map.get(b["equipment_id"])
        if eq:
            total += eq["cost_per_day"] * b["duration_days"]
    return total
